src/renderer/ducks/validator/sagas.ts

```diff
@@ -335,6 +341,87 @@ function* onBeaconNodeStatusChangeUpdateValidatorState(
     }
 }
 
+export function* getAttestationEffectiveness({
+    payload,
+    meta,
+}: ReturnType<typeof signedNewAttestation>): Generator<
+    SelectEffect | TakeEffect | AllEffect<CallEffect> | CallEffect,
+    void,
+    IValidatorComplete & ReturnType<typeof updateSlot> & (BlockAttestations[] | null)[]
+> {
+    const validator = yield select(getValidator, {publicKey: meta});
+    const config = getNetworkConfig(validator.network)?.eth2Config || mainnetConfig;
+
+    const eth2API = new CgEth2ApiClient(config, validator.beaconNodes[0]);
+
+    let lastSlot = payload.slot,
+        previousSlot = 0,
+        empty = 0,
+        skipped = 0,
+        skippedQue = 0,
+        inclusion = 0;
+    const timeoutSlot = config.params.SLOTS_PER_EPOCH + payload.slot;
+    while (true) {
+        const newSlot = yield take(updateSlot);
+        if (newSlot.meta !== validator.beaconNodes[0]) continue;
+        if (lastSlot >= newSlot.payload) continue;
+
+        // get all block information from last to current (mostly is only one but is important to get every block)
+        const range = new Array(newSlot.payload - lastSlot).fill(null).map((_, index) => lastSlot + index + 1);
+        const results = yield all(range.map((slot) => call(eth2API.beacon.blocks.getBlockAttestations, slot)));
+
+        results.forEach((result, index) => {
+            /** the main logic for collecting when is attestation may be included and the number of skipped blocks */
+            if (result === null) {
+                skippedQue++;
+            } else {
+                const sanitizedAttestations = result.filter(
+                    ({data}) => payload.block === data.beaconBlockRoot && data.index === payload.index,
+                );
+                if (!sanitizedAttestations.length) empty++;
+                else empty = 0;
+                sanitizedAttestations.forEach(({data}) => {
+                    if (data.slot > inclusion && data.slot > payload.slot) {
+                        inclusion = data.slot;
+                        skipped += skippedQue;
+                        skippedQue = 0;
+                    }
+                });
+                previousSlot = range[index];
+            }
+        });
+        /**
+         * break from loop after some number of blocks without information about followed attestations, whit that we
+         * can conclude that code fund block where is attestation included and there is no reason to continue this loop
+         * (in some rare case can cause premature breaking a loop, if is a problem increase "maxEmptyBlocks")
+         * */
+        const maxEmptyBlocks = 3;
```

Comment:
isn't that a bit low?

Reply:
think it is fine
if there is no watched attestation information we can stop watching,
for fail-safe, I set need to be 3 times in a row, and its rare case that there is no info about watched attestation in the block and then available in the next block
but I can increase it to 9 times

Reply:
Just to check if I understood correctly. You don't check if attestation distance is more than 3?

Reply:
yup

Reply:
tbh, I've seen distance up to 16 xD

Reply:
it is covered in the test :D
https://github.com/NodeFactoryIo/ChainGuardian/pull/416/files#diff-830870506b40d6475f8a3c8a8b0b68b5f467b54ebaed6c406a882bac49e70b42R177